### PR TITLE
Support for transcoding data from JSON to binary formats.

### PR DESCRIPTION
**Pull request by ryzhyk**
See commit messages for details.

### CODE REVIEW

File: test/datalog_tests/json_test.dl [hunk 1]
```diff
@@ -183,3 +183,16 @@ JsonTest(s32FromString1(),
          to_json_string_or_default(from_json_string(s32FromString1()): Result<S32FromString, string>)).
 JsonTest(s32FromString2(),
          to_json_string_or_default(from_json_string(s32FromString2()): Result<S32FromString, string>)).
+
+typedef WrappedEnum = JsonWrapper<TaggedEnum>
```
Comment (mbudiu-vmw) [inline test/datalog_tests/json_test.dl:187]: no reference output?

File: lib/json.dl [hunk 2]
```diff
@@ -20,7 +20,7 @@ extern function to_json_string(x: 'T): Result<string, string>
 
 /* Represents any valid JSON value.
  */
-#[rust="serde(from = \"serde_json::value::Value\", into = \"serde_json::value::Value\")"]
+#[rust="serde(from = \"__json::ValueWrapper\", into = \"__json::ValueWrapper\")"]
```
Comment (mbudiu-vmw) [inline lib/json.dl:23]: if the previous annotation was obscure, this one is doubly obscure.
Maybe once this API stabilizes we can wrap it into something that looks nicer for the user.

File: src/Language/DifferentialDatalog/Attribute.hs [hunk 3]
```diff
@@ -127,6 +134,23 @@ tdefCheckSizeAttr d TypeDef{..} =
          [Attribute{..}] -> err d attrPos $ "Invalid 'size' attribute: size must be an integer between 0 and " ++ show (maxBound::Int)
          _                    -> err d tdefPos $ "Multiple 'size' attributes are not allowed"
 
+{- 'custom_serde' attribute: Tells DDlog not to generate `Serialize` and
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Attribute.hs:137]: it is not useful to have this documentation just in the code, you have to surface it to the users as well.
there should be a section on annotations in the reference document?

File: lib/json.dl [hunk 4]
```diff
@@ -164,3 +164,27 @@ typedef JsonNum = // Integer number.  The value must be in the range between the
                 | // NaN and infinity are not valid JSON values.
                   // When serializing to JSON, such values will be replaced with 0.
                   JsonFloat{d: double}
+
+/* Wrapper type that serializes its inner object into binary formats via JSON string.
```
Comment (mbudiu-vmw) [inline lib/json.dl:168]: I don't really understand the connection between JSON, strings, and binary serialization.

File: test/datalog_tests/json_test.dl (hunk 1 above)
Comment (ryzhyk) [reply test/datalog_tests/json_test.dl:187]: Just added that.

File: lib/json.dl (hunk 2 above)
Comment (ryzhyk) [reply lib/json.dl:23]: Fortunately, this is internal to the library, users don't need to understand the annotations.

File: src/Language/DifferentialDatalog/Attribute.hs (hunk 3 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Attribute.hs:137]: There is a section in the tutorial. I'll add the new attribute there.

**Pull request merged**